src/core/project.js

```diff
@@ -44,13 +44,11 @@ const upgrades = {
   "0.3.1": upgradeFrom030,
 };
 
-export type ProjectJSON = Compatible<Project>;
-
-export function projectToJSON(p: Project): ProjectJSON {
+export function projectToJSON(p: Project): Compatible<Project> {
   return toCompat(COMPAT_INFO, p);
 }
 
-export function projectFromJSON(j: ProjectJSON): Project {
+export function projectFromJSON(j: Compatible<any>): Project {
```

Comment:
What if instead of throwing away the type information, we just enumerate all the supported Project types? E.g.

```js
type SupportedProject = Project_v_0_3_0 | Project_v_0_3_1 | Project_v_0_4_0;
type Project = Project_v_0_4_0 // latest;
export function projectToJSON(p: Project): Compatible<Project> {
export function projectFromJSON(j: Compatible<SupportedProject>): Project {
```

Reply:
I suppose that would help mostly for our upgrade functions. Reality is still we could get any json data as input at runtime. I'll have a look at doing this.

Reply:
@decentralion I've tried this approach, but an interesting issue with `opaque` came up.

https://github.com/sourcecred/sourcecred/commit/ffe296001439bdad746d801018d3ace47c161c70

Reply:
I really like it in ffe2960. Keeping all of the types around for the older serializations will make it **so much easier** for us to upgrade the types down the line and maintain backwards compatibility. Let's just remove opaque from compatible. I trust us to be responsible with the ability to construct arbitrary compatible objects in the codebase, so the opacity isn't really needed.

Reply:
I've included that commit :+1: